### PR TITLE
[TEVA-1018] Render job details notification box when there are multiple schools and create notification view component

### PR DESCRIPTION
**Pull request by josephhull676**
## Jira ticket URL

https://dfedigital.atlassian.net/browse/TEVA-1018

## Changes in this PR:

- Created view component for notification boxes
- Added new css selectors for notification box style-dependent background colours
- Render component if vacancy is associated with many organisations
- Added translations and changed old translations to fit into new structure

### Before

Screenshot 2020-09-02 at 16 55 20

### After

Screenshot 2020-09-02 at 16 56 58

### CODE REVIEW

File: app/components/jobseekers/multi_school_notification_component.rb [hunk 1]
```diff
@@ -0,0 +1,4 @@
+class Jobseekers::MultiSchoolNotificationComponent < ViewComponent::Base
```
Comment (cesidio) [inline app/components/jobseekers/multi_school_notification_component.rb:1]: You do not need a view component if there is no logic to handle

**Comment by cpjmcquillan** [comment]
The intention of this ticket (I think) was to create something that was configurable based on parameters passed in

**Comment by alexbowen** [comment]
Do you want to pair on this on monday @josephhull676 ?

File: app/views/shared/vacancy/_jobseeker_view.html.haml [hunk 2]
```diff
@@ -15,6 +15,9 @@
 
   .govuk-grid-column-two-thirds
 
+    - if @vacancy.organisations.many?
+      = render(Jobseekers::NotificationComponent.new(content: t('jobs.notifications.content.multi_school_job_notification_html')))
```
Comment (david-mears-dfe) [inline app/views/shared/vacancy/_jobseeker_view.html.haml:19]: I see Cesi has recommended using a partial instead, but it's worth knowing that a ViewComponent can be given a method `render?` which then lets you hide the decision whether or not to render, resulting in a cleaner, less logic-ful template. We've used it elsewhere in the app if you want to look at some examples.

@cesidio This might mean that there is in fact some benefit to using a ViewComponent here, because we remove 1 line of logic from the template. What do you think?

File: config/locales/en.yml [hunk 3]
```diff
@@ -241,9 +241,12 @@ en:
...
         other: '%{count} filters applied'
-    notification_labels:
-      new: new
-      action_required: action required
+    notifications:
+      notification_labels:
+        new: new
+        action_required: action required
+      content:
+        multi_school_job_notification_html: This teaching role is at <span class="govuk-!-font-weight-bold">more than one school in the trust</span>. You can find out more in Schools overview.
```
Comment (david-mears-dfe) [inline config/locales/en.yml:249]: Good non-flat structure here! https://ukgovernmentdfe.slack.com/archives/G0127R0KXQF/p1599140452052700

**Review by david-mears-dfe — APPROVED** [review]
This looks good to me, but note @cesidio 's comment. There is nothing wrong with your commit messages. They succinctly tell me what is in each of your succinct commits.

**Review by david-mears-dfe — CHANGES_REQUESTED** [review]
I am changing my review from 'approve' to 'request changes' as I'd like us either not to use a ViewComponent, or if we do, exploit the 'render?' helper method.

File: app/views/shared/vacancy/_jobseeker_view.html.haml (hunk 2 above)
Comment (david-mears-dfe) [reply app/views/shared/vacancy/_jobseeker_view.html.haml:19]: Now I remember @cpjmcquillan mentioned there was some complexity introduced by nesting components inside each other, but I don't know what complexity that is... care to comment?

File: app/components/notification_component.html.haml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+.govuk-notification.js-dismissable{role: 'alert', tabindex: '-1', class: notification_classes}
+  - if @dismiss
+    %a.govuk-link.govuk-link--no-visited-state.dismiss-link.js-dismissable__link{ href: '#' }
+      = t('buttons.dismiss')
+  - if render_title_and_body?
+    .govuk-notification__title= content[:title]
+    .govuk-notification__body= content[:body]
```
Comment (cpjmcquillan) [inline app/components/notification_component.html.haml:7]: `raw content[:body]`

File: app/components/notification_component.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+class NotificationComponent < ViewComponent::Base
+  def initialize(content:, style:, dismiss: true, background: false, alert: false)
+    @content = content
+    @style = style
+    @dismiss = style == 'danger' ? false : dismiss
+    @background = background
+    @alert = alert
```
Comment (cpjmcquillan) [inline app/components/notification_component.rb:7]: `%w[danger success].include?(style) ? false : alert` and add relevant specs

File: app/views/hiring_staff/sign_in/email/sessions/new.html.haml [hunk 6]
```diff
@@ -1,11 +1,7 @@
 - content_for :page_title_prefix, t('sign_in.title')
 
 .govuk-grid-row
-  .govuk-notification.govuk-notification--danger{ aria: { labelledby: "govuk-notification__title" }, role: "alert", tabindex: "-1" }
-    %h2.govuk-notification__title
-      =t('hiring_staff.temp_login.heading')
-    .govuk-notification__body
-      =t('hiring_staff.temp_login.please_use_email')
+  = render(NotificationComponent.new(content: { title: t('hiring_staff.temp_login.heading'), body: t('hiring_staff.temp_login.please_use_email') }, style: 'danger', dismiss: false))
```
Comment (cpjmcquillan) [inline app/views/hiring_staff/sign_in/email/sessions/new.html.haml:4]: No need to provide `dismiss: false` if style is `danger`

File: app/frontend/styles/components/flashes.scss [hunk 7]
```diff
@@ -68,6 +64,18 @@
   }
 }
 
+.govuk-notification--notice.govuk-notification__background {
```
Comment (david-mears-dfe) [inline app/frontend/styles/components/flashes.scss:67]: Having just searched, I am fairly sure we have no elements with the class `govuk-notification__background`, so we can get rid of these styles.

Comment (josephhull676) [reply app/frontend/styles/components/flashes.scss:67]: We do! The new notification box I've added for multi-school jobs, for example, has the following classes applied: alert govuk-notification govuk-notification--notice govuk-notification__background

File: spec/components/notification_component_spec.rb [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    end
+  end
+
+  context 'when background is false' do
+    it 'does not apply the background style' do
+      expect(inline_component.css('.govuk-notification__background')).to be_blank
+    end
+  end
+
+  context 'when style is success' do
```
Comment (cpjmcquillan) [inline spec/components/notification_component_spec.rb:83]: There is already a `when style is success` context, so you can move this spec there

File: spec/components/notification_component_spec.rb [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    let(:style) { 'success' }
+
+    it 'does not apply the alert style' do
+      expect(inline_component.css('.alert')).to be_blank
+    end
+  end
+
+  context 'when style is danger' do
```
Comment (cpjmcquillan) [inline spec/components/notification_component_spec.rb:91]: As above

File: spec/components/notification_component_spec.rb [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    let(:style) { 'danger' }
+
+    it 'does not apply the alert style' do
+      expect(inline_component.css('.alert')).to be_blank
+    end
+  end
+
+  context 'when style is notice' do
```
Comment (cpjmcquillan) [inline spec/components/notification_component_spec.rb:99]: As above

File: spec/components/notification_component_spec.rb (hunk 8 above)
Comment (josephhull676) [reply spec/components/notification_component_spec.rb:83]: How did I not notice that?...

Comment (cpjmcquillan) [reply spec/components/notification_component_spec.rb:83]: Je ne sais pas

**Review by david-mears-dfe — APPROVED** [review]

File: app/frontend/styles/components/flashes.scss (hunk 7 above)
Comment (david-mears-dfe) [reply app/frontend/styles/components/flashes.scss:67]: Ah ok!

File: spec/components/notification_component_spec.rb [hunk 11]
```diff
@@ -0,0 +1,112 @@
+require 'rails_helper'
+
+RSpec.describe NotificationComponent, type: :component do
+  let(:content) { 'This is content' }
+  let(:style) { 'notice' }
+  let(:dismiss) { true }
+  let(:background) { false }
+  let(:alert) { false }
+  let!(:inline_component) { render_inline(NotificationComponent.new(content: content, style: style, dismiss: dismiss, background: background, alert: alert)) }
```
Comment (david-mears-dfe) [inline spec/components/notification_component_spec.rb:9]: Any chance of this being split into multiple lines?

```ruby
  let!(:inline_component) { render_inline(NotificationComponent.new(
                                                                    content: content,
                                                                    style: style,
```
etc. You can let rubocop sort out the indentation.

File: spec/components/notification_component_spec.rb [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+  describe '#content' do
+    context 'when content is a string' do
+      it 'renders content in the body' do
+        expect(inline_component.css('.govuk-notification__body').to_html).to include(content)
```
Comment (david-mears-dfe) [inline spec/components/notification_component_spec.rb:14]: I would prefer to use `'This is content'` over `content` as it makes the test more robust (there's a small chance of a bug in the test making `content` evaluate to `nil`, for example). Up to you though.

File: spec/components/notification_component_spec.rb [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+      end
+    end
+
+    context 'when content is a hash' do
+      let(:content) { { title: 'Title', body: 'This is the body' } }
+
+      it 'renders the title' do
+        expect(inline_component.css('.govuk-notification__title').to_html).to include(content[:title])
```
Comment (david-mears-dfe) [inline spec/components/notification_component_spec.rb:22]: (Likewise, I would have preferred 'Title' over content[:title])

File: spec/components/notification_component_spec.rb [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+    context 'when dismiss is false' do
+      let(:dismiss) { false }
+
+      it 'does not render the dismiss link' do
+        expect(rendered_component).to_not include(I18n.t('buttons.dismiss'))
+      end
+    end
+  end
+
+  describe '#dismiss' do
```
Comment (david-mears-dfe) [inline spec/components/notification_component_spec.rb:47]: Not sure you meant that this describes #dismiss? It looks like it describes style.

**Review by david-mears-dfe — COMMENTED** [review]
Nice. I left a few more comments but they're not deal breakers.

**Review by alexbowen — APPROVED** [review]
looks good to me

File: spec/components/notification_component_spec.rb (hunk 14 above)
Comment (josephhull676) [reply spec/components/notification_component_spec.rb:47]: Yep. Mistake I've now fixed.

File: spec/components/notification_component_spec.rb (hunk 13 above)
Comment (josephhull676) [reply spec/components/notification_component_spec.rb:22]: Done!

File: spec/components/notification_component_spec.rb (hunk 12 above)
Comment (josephhull676) [reply spec/components/notification_component_spec.rb:14]: Done!

File: spec/components/notification_component_spec.rb (hunk 11 above)
Comment (josephhull676) [reply spec/components/notification_component_spec.rb:9]: Done!

File: spec/components/notification_component_spec.rb [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+    context 'when style is success' do
+      let(:style) { 'success' }
+      let(:alert) { 'true' }
+
+      it 'applies correct class' do
+        expect(inline_component.css('.govuk-notification--success')).to_not be_blank
+      end
+
+      it 'removes the apply class' do
```
Comment (cpjmcquillan) [inline spec/components/notification_component_spec.rb:84]: `it 'does not apply the alert class'`

File: spec/components/notification_component_spec.rb [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+      context 'when alert is false' do
+        it 'does not apply the alert class' do
+          expect(inline_component.css('.alert')).to be_blank
+        end
+      end
+    end
+
+    context 'when style is success' do
+      let(:style) { 'success' }
+      let(:alert) { 'true' }
```
Comment (cpjmcquillan) [inline spec/components/notification_component_spec.rb:78]: Use a boolean not a string

File: spec/components/notification_component_spec.rb [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+      end
+
+      it 'removes the apply class' do
+        expect(inline_component.css('.alert')).to be_blank
+      end
+    end
+
+    context 'when style is danger' do
+      let(:style) { 'danger' }
+      let(:alert) { 'true' }
```
Comment (cpjmcquillan) [inline spec/components/notification_component_spec.rb:91]: Use boolean instead of a string

File: spec/components/notification_component_spec.rb [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+
+      it 'does not render the dismiss link' do
+        expect(rendered_component).to_not include(I18n.t('buttons.dismiss'))
+      end
+
+      it 'applies correct class' do
+        expect(inline_component.css('.govuk-notification--danger')).to_not be_blank
+      end
+
+      it 'removes the apply class' do
```
Comment (cpjmcquillan) [inline spec/components/notification_component_spec.rb:101]: `it 'does not apply the alert class'`

**Review by cpjmcquillan — APPROVED** [review]
Nice work 👍
